<?xml version="1.0"?>
<robot name="quadrotor" xmlns:xacro="http://www.ros.org/wiki/xacro">

    <!-- Arguments -->
    <xacro:arg name="namespace" default="NX01" />
    <xacro:arg name="d435_range_max_depth" default="5.0" />

    <!-- Constants -->
    <!-- <xacro:property name="M_PI" value="3.1415926535897931" /> -->

    <!-- Included URDF Files -->
    <xacro:include filename="$(find mighty)/urdf/quadrotor_base.urdf.xacro" />
    
    <!-- Instantiate quadrotor_base_macro once (has no parameters atm) -->
    <xacro:quadrotor_base_macro />

    <!-- Note: added $(arg namespace)_ to the name to avoid having multiple sensors with the same name when using multiple agents -->

    <!-- Livox -->
    <xacro:include filename="$(find ros2_livox_simulation)/urdf/mid360.xacro" />
    <xacro:mid360 name="$(arg namespace)_livox" namespace="$(arg namespace)" parent="base_link" topic="mid360" >
      <origin xyz="0 0 0.2" rpy="0 0.3 0"/>
    </xacro:mid360>

    <!-- Forward facing camera (d435)-->
    <xacro:include filename="$(find mighty)/urdf/_d435.urdf.xacro" />
    <xacro:sensor_d435 name="$(arg namespace)_d435" topics_ns="$(arg namespace)" range_max_depth="$(arg d435_range_max_depth)" parent="base_link">
      <origin xyz="0.05 0.0 0" rpy="0 0 0"/>
    </xacro:sensor_d435>
    
    <!-- Forward facing camera (d435i)-->
    <!-- <xacro:include filename="$(find mighty)/urdf/_d435i.urdf.xacro" />
    <xacro:sensor_d435i name="$(arg namespace)_d435i" topics_ns="$(arg namespace)" parent="base_link">
      <origin xyz="0.2 0.0 0" rpy="0 0 0"/>
    </xacro:sensor_d435i> -->

    <!-- Forward facing camera -->
    <!-- TODO: Not sure why but can't put them in namespace  -->
    <!-- <xacro:include filename="$(find mighty)/urdf/generic_camera.urdf.xacro" />
    <xacro:generic_camera name="$(arg namespace)_front_cam" parent="base_link" ros_topic="$(arg namespace)" cam_info_topic="$(arg namespace)" update_rate="50" res_x="320" res_y="240" image_format="R8G8B8" hfov="90">
      <origin xyz="0.1 0.0 -0.1" rpy="0 0 0"/>
    </xacro:generic_camera> -->

    <!-- Forward facing camera (t265)-->
    <!-- <xacro:include filename="$(find mighty)/urdf/_t265.urdf.xacro" />
    <xacro:sensor_t265 name="front_cam" topics_ns="$(arg namespace)" parent="base_link">
      <origin xyz="0.1 0.0 0.1" rpy="0 0 0"/>
    </xacro:sensor_t265> -->
    
    <!-- Sonar height sensor -->
    <!-- <xacro:include filename="$(find mighty)/urdf/sonar_sensor.urdf.xacro" /> 
    <xacro:sonar_sensor name="sonar" parent="base_link" >
      <origin xyz="-0.03 -0.03 1.0" rpy="0 ${90*pi/180} 0"/>
    </xacro:sonar_sensor> -->

    <!-- Forward-facing spotlight -->
    <xacro:macro name="spotlight_macro" params="namespace parent">
      <gazebo reference="${parent}">
        <!-- <light name="${namespace}_spotlight0" type="spot">
          <pose>0 -0.1 -0.05 0 -1.0472 0.3</pose>
          <diffuse>30 30 30 1</diffuse>
          <specular>1 1 1 1</specular>
          <direction>0 0 1</direction>
          <attenuation>
            <range>30</range>
            <constant>0.05</constant>
            <linear>0.005</linear>
            <quadratic>0.001</quadratic>
          </attenuation>
          <spot>
            <inner_angle>20</inner_angle>
            <outer_angle>30</outer_angle>
            <falloff>1</falloff>
          </spot>
          <cast_shadows>true</cast_shadows>
        </light> -->

        <light name="${namespace}_spotlight1" type="spot">
          <pose>0 0 -0.05 0 -1.0472 0</pose>
          <diffuse>30 30 30 1</diffuse>
          <specular>1 1 1 1</specular>
          <direction>0 0 1</direction>
          <attenuation>
            <range>30</range>
            <constant>0.05</constant>
            <linear>0.005</linear>
            <quadratic>0.001</quadratic>
          </attenuation>
          <spot>
            <inner_angle>20</inner_angle>
            <outer_angle>30</outer_angle>
            <falloff>1</falloff>
          </spot>
          <cast_shadows>true</cast_shadows>
        </light>

        <!-- <light name="${namespace}_spotlight2" type="spot">
          <pose>0 0.1 -0.05 0 -1.0472 -0.3</pose>
          <diffuse>30 30 30 1</diffuse>
          <specular>1 1 1 1</specular>
          <direction>0 0 1</direction>
          <attenuation>
            <range>30</range>
            <constant>0.05</constant>
            <linear>0.005</linear>
            <quadratic>0.001</quadratic>
          </attenuation>
          <spot>
            <inner_angle>20</inner_angle>
            <outer_angle>30</outer_angle>
            <falloff>1</falloff>
          </spot>
          <cast_shadows>true</cast_shadows>
        </light> -->
      </gazebo>
    </xacro:macro>

    <!-- <xacro:spotlight_macro namespace="$(arg namespace)" parent="base_link"/> -->

</robot>
